ament_mypy/ament_mypy/main.py

```diff
@@ -0,0 +1,246 @@
+        default=[os.curdir],
+        help='The files or directories to check. For directories files ending '
+             "in '.py' will be considered."
+    )
+    parser.add_argument(
+        '--exclude',
+        metavar='filename',
+        nargs='*',
+        dest='excludes',
+        help='The filenames to exclude.'
+    )
+    parser.add_argument(
+        '--cache-dir',
+        metavar='cache',
+        default=os.devnull,
+        dest='cache_dir',
+        help='The location mypy will place its cache in. Defaults to system '
+             'null device'
+    )
+
+    # not using a file handle directly
+    # in order to prevent leaving an empty file when something fails early
+    parser.add_argument(
+        '--xunit-file',
+        help='Generate a xunit compliant XML file'
+    )
+    args = parser.parse_args(argv)
+
+    if args.xunit_file:
+        start_time = time.time()
+
+    if args.config_file and not os.path.exists(args.config_file):
+        print("Could not find config file '{}'".format(args.config_file), file=sys.stderr)
+        return 1
+
+    filenames = _get_files(args.paths)
+    if args.excludes:
+        print('excluding: ', args.excludes)
+        filenames = [f for f in filenames
+                     if os.path.basename(f) not in args.excludes]
+    if not filenames:
+        print('No files found', file=sys.stderr)
+        return 1
+
+    normal_report, error_messages, exit_code = _generate_mypy_report(
+        filenames,
+        args.config_file,
+        args.cache_dir
+    )
+
+    if error_messages:
+        print('mypy error encountered', file=sys.stderr)
+        print(error_messages, file=sys.stderr)
+        print('\nRegular report continues:')
+        print(normal_report, file=sys.stderr)
+        return exit_code
+
+    errors_parsed = _get_errors(normal_report)
+
+    print('\n{} files checked'.format(len(filenames)))
+    if not normal_report:
+        print('No errors found')
+    else:
+        print('{} errors'.format(len(errors_parsed)))
+
+    print(normal_report)
+
+    print('\nChecked files:')
+    print(''.join(['\n* {}'.format(f) for f in filenames]))
+
+    # generate xunit file
+    if args.xunit_file:
+        folder_name = os.path.basename(os.path.dirname(args.xunit_file))
+        file_name = os.path.basename(args.xunit_file)
+        suffix = '.xml'
+        if file_name.endswith(suffix):
+            file_name = file_name[:-len(suffix)]
+            suffix = '.xunit'
+            if file_name.endswith(suffix):
+                file_name = file_name[:-len(suffix)]
+        testname = '{}.{}'.format(folder_name, file_name)
+
+        xml = _get_xunit_content(errors_parsed, testname, filenames, time.time() - start_time)
+        path = os.path.dirname(os.path.abspath(args.xunit_file))
+        if not os.path.exists(path):
+            os.makedirs(path)
+        with open(args.xunit_file, 'w') as f:
+            f.write(xml)
+
+    return 1 if normal_report else 0
```

Comment:
By not using the exit code here you're masking actual failures. Please try installing mypy from pip and setting MYPYPATH = PYTHONPATH and notice that this silently passes since mypy crashed. Such a case should be covered by a test, please.

Reply:
7ef37f1 fixed